ADIOS1 backend

anokfireball opened the pull request:
This adds both a serial and parallel ADIOS1 backend.

One potential drawback of this backend is that it can only mutually exclusively build a non-MPI/parallel version. This might induce contraints on other backends as well, depending on how they are built.

The implementation in this PR does the bare minimum to use ADIOS1 effectively:
 - It considers the typical ADIOS workflow of defining variables before opening a file and tries to reduce the number of times a file has to be (re-)opened. To do this, file opening is deferred until the first write must happen. Two IO queues are used in the backend to discern the two phases of operations (before / after file open). One condition under which re-opening a file can not be avoided is when a manual ```Series::flush()``` is called before new a variable is defined (there may be more complex ones that we have not recognized yet).
 - It bunches reads for every file to schedule a combined read of multiple chunks.
 - It keeps file handles open as long as is possible. It also reopenes them when required.
 - It allows to write dynamically sized chunks per dataset. Typically, ADIOS recommends to define a unique variable for every different-sized/-localized chunk for every dataset. We solve this by creating scalar varaibles for every every dimension of the extent and offset (in "/tmp", so it does not conflict with the openPMD standard). After being defined once, these variables are re-used as the extent and offset for any chunk of a dataset. To do this, the scalar variables just have to be overwritten with the desired extent and offset values before performing the chunk write.

anokfireball committed:
Build Skeleton of ADIOS1 backend

ax3l commented on CMakeLists.txt:
```diff
@@ -41,7 +41,7 @@ endfunction()
 openpmd_option(MPI    "Enable MPI support"     AUTO)
 openpmd_option(HDF5   "Enable HDF5 support"    AUTO)
-openpmd_option(ADIOS1 "Enable ADIOS1 support"  OFF)
+openpmd_option(ADIOS1 "Enable ADIOS1 support"  AUTO)
```
also occurs in `README.md`, `docs/source/dev/buildoptions.rst` and `docs/source/dev/dependencies.rst`

anokfireball committed:
Get desired ADIOS1 version through CMake

Determine and check the parallel/serial status of a found ADIOS1 lib

anokfireball committed:
ADIOS1 path & dataset creation (untested)

ax3l commented on CMakeLists.txt:
```diff
@@ -158,11 +158,15 @@ else()
-#endif()
+# TODO we could support more combinations than MPI+pADIOS and noMPI+sADIOS
+if(openPMD_HAVE_MPI AND openPMD_HAVE_ADIOS1 AND ADIOS_HAVE_SEQUENTIAL)
+    message(FATAL_ERROR "Found MPI but requested ADIOS1 is serial. "
```
are you sure a MPI-parallel ADIOS not not still ship a "sequential" component "as well"?

anokfireball replied:
The library itself might provide both, but with the used ```FindAdios.cmake``` only mutually exclusive versions get linked. Since ADIOS has a unified interface (with [dummy MPI](https://github.com/ornladios/ADIOS/blob/master/src/public/adios_mpi.h) if serial), I doubt you can link both simultaneously.

ax3l commented on cmake/FindADIOS.cmake:
```diff
@@ -178,6 +179,9 @@ if(ADIOS_FOUND)
     endforeach()
     # we could append ${CMAKE_PREFIX_PATH} now but that is not really necessary
 
+    # determine whether found library is serial only
```
I think we can only determine:
- is serial *and* parallel
- is serial only

anokfireball replied:
Alright, the vocabulary in this comment is not sufficient. This determines if the library *links*  as serial only with the specified components.
The library itself might be serial + parallel.

ax3l replied:
ok cool.

want I wanted to say is just: if you build serial, the `lib/` dir will contain a regular and a `_nompi` version, as far as I remember.

If you check that only the noMPI is linked this is fine, but you might want to prefer it someway during search and validate it with two adios installations (/w and /wo MPI), printing all vars to make sure. Also includes (and their `#defines`) should match.

anokfireball committed:
Serial ADIOS1 integration

Bump required ADIOS1 version to 1.13.0.
Enable serial ADIOS1 through -D_NOMPI.
Enable tests according to used backends.

anokfireball committed:
ADIOS1 attribute writing

Move ADIOS1 logic form parallel to serial (dummy MPI allows for mostly
the same code).

ax3l commented on .travis/spack/packages.yaml:
```diff
@@ -16,6 +16,9 @@ packages:
       openmpi@1.6.5%gcc@7.2.0 arch=linux-ubuntu14-x86_64: /usr
       openmpi@1.6.5%clang@5.0.0 arch=linux-ubuntu14-x86_64: /usr
     buildable: False
+packages:
```
oh oh, pls remove this single line

ax3l commented:
I think there is more to do for the CI than we can do in the next few days.
Just leave it off for now (remove the last three commits) and we will add travis for ADIOS in a follow-up together end of the week.

anokfireball commented:
This is probably not a problem with our build process. The errors are caused by timeouts during download of our dependencies (e.g. https://zlib.net was down yesterday). Now, Boost & HDF5 are horribly slow, but making progress.

ax3l commented:
Yes, also that but I restarted them later on and saw some other issues, also spack-dependency-wise, that I need to investigate properly.

anokfireball committed:
ADIOS dataset/chunk writing

anokfireball committed:
Adapt error messages to context

anokfireball committed:
Update documentation

Remove not yet implemented status for ADIOS1.
Bump required ADIOS1 version.

Commit:
Merge branch 'dev' into topic/adios

ax3l commented on include/openPMD/auxiliary/Memory.hpp:
```diff
@@ -83,6 +99,9 @@ allocatePtr(Datatype dtype, size_t numPoints)
             del = [](void* p){ delete[] static_cast< bool* >(p); p = nullptr; };
             break;
+        case DT::STRING:
+            /* user assings c_str pointer */
```
typo: assigns

ax3l commented on src/IO/ADIOS/ADIOS1IOHandler.cpp:
```diff
@@ -37,31 +95,527 @@ ADIOS1IOHandler::flush()
+                          dims.c_str(),
+                          global_dims.c_str(),
+                          local_offsets.c_str());
+    ASSERT(id >= 0 /* ??? */, "Internal error: Failed to define ADIOS variable during Dataset writing");
```
good candidate for a `VERIFY` macro - seen this a lot in the last on startup/env issues.

anokfireball replied:
Very much agreed. We should not exclude these assertions in production code, unless the user *really* wants to.

ax3l commented on src/IO/ADIOS/ADIOS1IOHandler.cpp:
```diff
@@ -37,31 +95,527 @@ ADIOS1IOHandler::flush()
+            break;
+        }
+        case DT::BOOL:
+            throw std::runtime_error("No workaround for ADIOS1 bool implemented");
```
Current PIConGPU representation: https://github.com/ComputationalRadiationPhysics/picongpu/pull/1756

also just as ints, unfortunately.

anokfireball replied:
Currently tinkering with this. Adapting the behaviour of using unsigned bytes for bools means we can store them but not read them back typesafe.
Having every uint8_t of value 0 be reinterpreted to a bool false and every uint8_t of value 1 be reinterpreted to a bool true is undesirable when reading a file.

anokfireball committed:
ADIOS attribute reading

anokfireball committed:
Merge branch 'topic/adios' of github.com:C0nsultant/libopenPMD into topic/adios

anokfireball committed:
ADIOS attribute listing, path opening

anokfireball committed:
ADIOS file reading

Dataset opening, path & dataset listing.
Change frontend logic to rely less on HDF5 backend implementation
quirks.
Working ADIOS test case.

anokfireball committed:
ADIOS dataset reading

Working ADIOS read test case.
ADIOS attribute datatype test case.

anokfireball committed:
Parallel ADIOS

Working parallel reads.
This only does MPI inside ADIOS, not the desired MPI_AGGREGATE.

anokfireball committed:
Parallel ADIOS write test

anokfireball committed:
Dynamic ADIOS chunk writing

Sidestep repeated file opening & closing when writing chunks by having
scalar variables for every dimension. These values are re-used as the
extent & offset of a chunk.

anokfireball committed:
Reorder ADIOS IO operations

To fit ADIOS's workflow better and to avoid redundant file open/file
closing, separate the IO operations in ADIOS1 backend into two queues:

Defines and file context preperation (m_setup) that gets drained before
a file is opened for writing.

Actual data transfers that require an open file but do not re-open it (
m_work).

anokfireball committed:
Bump required ADIOS version

1.13.1 is very mich desired.

anokfireball committed:
Copy behaviour to Parallel ADIOS1

anokfireball committed:
Fix compilation when ADIOS is missing

anokfireball committed:
Switch to MPI_AGGREGATE

anokfireball committed:
ADIOS test cases

ax3l commented on include/openPMD/auxiliary/StringManip.hpp:
```diff
@@ -126,8 +126,9 @@ join(std::vector< std::string > const& vs, std::string const& delimiter)
             std::ostringstream ss;
             std::copy(vs.begin(),
-                      vs.end(),
+                      vs.end() - 1,
```
o.0

this looks dirty and unsave for zero-lengths

ax3l commented on include/openPMD/auxiliary/StringManip.hpp:
```diff
@@ -126,8 +126,9 @@ join(std::vector< std::string > const& vs, std::string const& delimiter)
             std::ostringstream ss;
             std::copy(vs.begin(),
-                      vs.end(),
+                      vs.end() - 1,
```
o.0

this changes looks a bit unsave, e.g. for zero-lengths :)

ax3l replied:
offline discussion: add comment `// do not append separator after last element`

anokfireball replied:
The empty and size 1 vector cases are handled above ;)

ax3l replied:
yep, saw this too late :)

ax3l commented:
@C0nsultant can you pls add a little to-do list in the PR description what is still missing in this PR? :) Just curious :)

anokfireball commented:
I added the major two things that have me refrain from merging yet.
Nothing major, at least the testing can be done in a few hours (should have  time for that tomorrow).

anokfireball committed:
Fix ADIOS fileBased Series logic

Adapt the different ADIOS workflow to work with fileBased output. This
requires closing and flushing open handles as ADIOS handles are only
one-directional (in- or out-put).

ax3l commented:
sounds good.
leave the CI out for now, it looks still buggy and I had no time to progress on it.

anokfireball committed:
ADIOS bool workaround

Treat booleans as unsigned bytes. This allows us to write them, but not
read them back as bool (they are uint8_t).

anokfireball committed:
Use Tranform parameters on Datasets

anokfireball committed:
Correct error text

anokfireball committed:
Update Howto backend documentation

Include new matcher functionality in tutorial.

anokfireball committed:
Add non-functional ADIOS backend commands

ADIOS1 does not offer deletion inside files, so handle those request
non-gracefully.

Commit:
Merge branch 'dev' into topic/adios

ax3l commented:
> can only mutually exclusively build a serial/parallel version

@C0nsultant but if I build the parallel version and satisfy the MPI dependency e.g. when running in mpiexec - can I still create a serial `Series` from a single rank? This should work, no?

ax3l commented on CMakeLists.txt:
```diff
@@ -301,6 +303,10 @@ if(openPMD_HAVE_ADIOS1)
     target_include_directories(openPMD SYSTEM PUBLIC ${ADIOS_INCLUDE_DIRS})
     target_compile_definitions(openPMD PUBLIC "-DopenPMD_HAVE_ADIOS1=1")
+    if(ADIOS_HAVE_SEQUENTIAL)
+        # TODO might be smarter to get ALL definitions from adios-config -s & parse it
```
Yes, ideally in `FindADIOS.cmake` we should populate `ADIOS_DEFINITIONS`. We then use that var here.

ax3l commented on CMakeLists.txt:
```diff
@@ -399,17 +405,20 @@ if(BUILD_TESTING)
-    if(${examplename} MATCHES ".+parallel$")
-        if(openPMD_HAVE_MPI)
+# Current examples all use HDF5, elaborate if other backends are used
+if(openPMD_HAVE_HDF5)
```
don't exclude the build, just the test below is fine.
the examples should still be build and only throw errors at runtime if they try to use a backend that does not exist.

anokfireball replied:
I included the examples as tests in CI. With this approach, CI tests without HDF5 will always fail.

ax3l replied:
no, the `add_tests` are in the block below and properly guarded. Those are the runtimeand require HDF5 files and backend. but the compile should still be done.

ax3l commented on include/openPMD/IO/ADIOS/ADIOS1Auxiliary.hpp:
```diff
@@ -0,0 +1,150 @@
+        case DT::UNDEFINED:
+            throw std::runtime_error("Unknown Attribute datatype");
+        default:
+            throw std::runtime_error("Datatype not implemented in HDF5 IO");
```
ADIOS :)

ax3l commented on include/openPMD/Series.hpp:
```diff
@@ -279,4 +280,12 @@ Format determineFormat(std::string const& filename);
+/** Create a functor to determine if a file can be of a format given the filename on disk.
+ *
+ * @param   name        String containing desired filename without filename extension.
+ * @param   f           File format to check plausibility for.
```
Maybe write "backend applicability" instead of "plausibility"?

ax3l commented on src/IO/ADIOS/ADIOS1IOHandler.cpp:
```diff
@@ -53,15 +240,1228 @@ ADIOS1IOHandler::flush()
+            break;
+        case DT::UNDEFINED:
+        case DT::DATATYPE:
+            throw std::runtime_error("Unknown Attribute datatype");
```
I am wondering if one has still to `break` after a throw...?
probably, no?

anokfireball replied:
Execution at that point aborts until you reach the next mathcing ```catch``` (if it reaches the top level, the process aborts). And even inside that catch, you can not possilbly recover to the point of failure (unless you use ```goto```-fuckery).

So no, there is no ```break``` required after a ```throw``` just like after a ```return```.

ax3l replied:
ok, I was more wondering if the `default` can be reached without the `break`.

ax3l commented on src/IO/ADIOS/ADIOS1IOHandler.cpp:
```diff
@@ -53,15 +240,1228 @@ ADIOS1IOHandler::flush()
+                for( int i = 0; i < size; ++i )
+                {
+                    vs[i] = auxiliary::strip(std::string(c[i], std::strlen(c[i])), {'\0'});
+                    /* TODO pointer should be freed, but this causes memory curruption */
```
make this a doxygen todo please: `/** @todo ...`

ax3l commented on src/IO/ADIOS/ParallelADIOS1IOHandler.cpp:
```diff
@@ -20,36 +20,243 @@
+    char const* c = params.str().c_str();
+
+    int status;
+    /* TODO ADIOS_READ_METHOD_BP_AGGREGATE */
```
`//! @todo ...`

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -1087,10 +1089,407 @@ TEST_CASE( "no_serial_hdf5", "[serial][hdf5]" )
+    REQUIRE(s.getAttribute("vecDouble").get< std::vector< double > >() == std::vector< double >({0., 1.79769e+308}));
+    REQUIRE(s.getAttribute("vecLongdouble").get< std::vector< long double > >() == std::vector< long double >({0.L, 1.18973e+4932L}));
+    REQUIRE(s.getAttribute("vecString").get< std::vector< std::string > >() == std::vector< std::string >({"vector", "of", "strings"}));
+    REQUIRE(s.getAttribute("bool").get< uint8_t >() == static_cast< uint8_t >(true));
```
can you add a comment on bool reads please?
maybe we also want to write this in the manual in a section on "backends: features & limitations" or so.

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -1087,10 +1089,407 @@ TEST_CASE( "no_serial_hdf5", "[serial][hdf5]" )
+
+TEST_CASE( "hzdr_adios1_sample_content_test", "[serial][adios1]" )
+{
+    // since this file might not be publicly available, gracefully handle errors
```
please add `/** @todo add bp example files to https://github.com/openPMD/openPMD-example-datasets */`

anokfireball commented:
> but if I build the parallel version and satisfy the MPI dependency e.g. when running in mpiexec - can I still create a serial Series from a single rank? This should work, no?

Yes. Yes, that always works with single-size MPI communicators.
Let's rephrase that to non-MPI and parallel versions.

anokfireball committed:
Parse ADIOS compile definitions

Use adios_config -c to get compile definitions.

anokfireball committed:
Cleanup

anokfireball committed:
Merge branch 'topic/adios' of github.com:C0nsultant/libopenPMD into topic/adios

ax3l commented:
Funny, I just saw this: https://github.com/ornladios/ADIOS/blob/v1.13.1/KNOWN_BUGS

That means writing a var for each chunk's dimensions is indeed the right way to go in ADIOS1 ;)
